### PR TITLE
Main Branch Merge

### PR DESCRIPTION
**Pull request by obj809**
Merging main-development branch into main branch.

### CODE REVIEW

File: js/calculatorApp.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        } else if (operator === "÷") {
+            if (num2 === 0) {
+                return "Cannot divide by zero";
+            } else {
+                return num1 / num2;
+            }
+        } else {
+            return "Invalid operator";
+        }
+    };
```
Comment (martynakr) [inline js/calculatorApp.js:30]: take this out of initializeCalcualtor function

File: js/calculatorApp.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+
+export const initializeCalculator = () => {
+    const buttonArr = ['AC', '+/-', '%', '÷', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];
```
Comment (martynakr) [inline js/calculatorApp.js:3]: Good idea to have a function like this but I would seperate the different things that are happening here:
1. initializeCalculator should create elements for the calculator and add calculator where it should go on the page
2. calculate - just take it out of this function
3. Select all the calcualtor button and add event listeners to all of them outsside of here:
```js
const btns = document.getElementsByClassname("calc-btn")
```

Loop through btns

File: js/calculatorApp.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                return "Cannot divide by zero";
+            } else {
+                return num1 / num2;
+            }
+        } else {
+            return "Invalid operator";
+        }
+    };
+
+    let createRowAndButtons = () => {
```
Comment (martynakr) [inline js/calculatorApp.js:32]: when creating functions, use const so you don't accidentally override to a different data type

File: js/calculatorApp.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        let index = 0;
+        for (let i = 0; i < numberOfRows; i++) {
+            const newCalculatorRow = document.createElement('div');
+            calculatorBody.appendChild(newCalculatorRow);
+            const buttonsInRow = Math.min(buttonArr.length - index, 4);
+
+            for (let j = 0; j < buttonsInRow; j++) {
+                const newButton = document.createElement('button');
+                newButton.innerText = buttonArr[index];
+                newCalculatorRow.appendChild(newButton);
```
Comment (martynakr) [inline js/calculatorApp.js:42]:
```js
newButton.classList.add('calc-btn')
```

File: js/calculatorApp.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                        calculatorInput.value = num1;
+                    } else if (newButton.innerText === '=') {
+                        num2 = parseFloat(calculatorInput.value);
+                        calculatorInput.value = calculate(num1, num2, operator);
+                    } else {
+                        calculatorInput.value = '';
+                        calculatorInput.value += newButton.innerText;
+                    }
+                });
+                index += 1;
```
Comment (martynakr) [inline js/calculatorApp.js:61]: this should be outside of the function that creates calculator elements

File: js/calculatorApp.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    const calculate = (num1, num2, operator) => {
+        if (operator === "+") {
+            return num1 + num2;
+        } else if (operator === "-") {
+            return num1 - num2;
+        } else if (operator === "x") {
+            return num1 * num2;
+        } else if (operator === "÷") {
+            if (num2 === 0) {
+                return "Cannot divide by zero";
```
Comment (martynakr) [inline js/calculatorApp.js:23]: throw errors here so it's more semantically correct

File: js/calculatorApp.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                });
+                index += 1;
+            }
+        }
+    };
+
+    createRowAndButtons();
+}
+
+document.getElementById('calculatorCard').addEventListener('dblclick', function() {
```
Comment (martynakr) [inline js/calculatorApp.js:69]: replace with arrow functions here

File: js/notesApp.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            if (taskInput.value.trim() === '') return; 
+    
+            const li = document.createElement('li');
+            const taskContent = document.createElement('span');
+            taskContent.textContent = taskInput.value;
+    
+            const removeButton = document.createElement('button');
+            removeButton.textContent = 'Remove';
+            removeButton.className = 'removeTask';
+            removeButton.onclick = function() {
```
Comment (martynakr) [inline js/notesApp.js:47]: try and no use onclick but stick to addEventListener

File: js/notesApp.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        const tasksList = document.createElement('ul');
+        tasksList.setAttribute('id', 'tasksList');
+    
+        contentWrapper.appendChild(heading);
+        contentWrapper.appendChild(inputContainer);
+        contentWrapper.appendChild(tasksList);
+    
+        addTaskBtn.addEventListener('click', addTask);
+    
+        function addTask() {
```
Comment (martynakr) [inline js/notesApp.js:37]: add the value of input as a parameter to this function

File: js/notesApp.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    
+        contentWrapper.innerHTML = '';
+    
+        const heading = document.createElement('h2');
+        heading.textContent = 'Task Tracker App';
+    
+        const inputContainer = document.createElement('div');
+        inputContainer.setAttribute('id', 'inputContainer');
+        
+        const taskInput = document.createElement('input');
```
Comment (martynakr) [inline js/notesApp.js:16]:
```js
const attributes = [{name: "id", value: "taskInput"}]

const createDomElement = (elType, attributes, textContent) => {
    const el = document.createElement(elType)
    attributes.forEach((att) => {
       el.setAttribute(att.name, att.value)
    })
    
    if(textContent) {
       el.textContent = textContent
    }
  return el
}
```

File: js/timeModule.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+
+export const currentTime = () => {
+    let now = new Date();
+    let hours = now.getHours();
+    let minutes = now.getMinutes();
+  
+    minutes = minutes < 10 ? '0' + minutes : minutes;
+  
+    let formattedTime = hours + ':' + minutes;
```
Comment (martynakr) [inline js/timeModule.js:9]: Have a look at [toLocaleTimeString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString)

File: js/weatherApp.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
+export const initializeWeatherApp = () => {
+    document.getElementById('weatherCard').addEventListener('dblclick', function() {
+        document.getElementById('weatherModal').showModal();
+        const contentWrapper = document.querySelector('#weatherModal .content-wrapper');
+        contentWrapper.innerHTML = ''; 
+
+        const weatherAppHtml = `
+            <h2>Javascript Weather App</h2>
```
Comment (martynakr) [inline js/weatherApp.js:8]: If you had a general createDOomELement function you could replace this with calling that function

File: js/weatherApp.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                    <p>Wind Speed: ${weatherData.wind.speed} m/s</p>
+                </div>
+            `;
+
+            container.innerHTML = cardHtml;
+        };
+
+        document.getElementById('getWeatherButtonModal').addEventListener('click', async function() {
+            const city = document.getElementById('cityInput').value;
+            const apiKey = '448f568c5483d5b0f66f39e0a3ac8fbc';
```
Comment (martynakr) [inline js/weatherApp.js:35]: don't expose this on github, create a .env file and import from there, also add that file to .gitignore

File: js/weatherApp.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+                    const weatherData = await weatherResponse.json();
+
+                    window.displayWeatherCard(weatherData);
+                } else {
+                    alert('City not found!');
+                    console.error('City not found');
+                }
+            } catch (error) {
+                console.error('Error fetching data:', error);
+            }
```
Comment (martynakr) [inline js/weatherApp.js:56]: this should be its own function 
```js
const getLatLong = (city) => {
 const geoResponse = await fetch(`http://api.openweathermap.org/geo/1.0/direct?q=${city}&limit=1&appid=${apiKey}`;
  const geoData = await geoResponse.json();
  if (geoData.length < 0) {
     throw new Error("")
     }
      
 return geoData;

}
```

Comment (martynakr) [reply js/weatherApp.js:56]:
```js
const getWeatherData = (geoData) => {
   const { lat, lon } = geoData[0];
   const weatherUrl = `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}`;

       const weatherResponse = await fetch(weatherUrl);
       const weatherData = await weatherResponse.json();
       return weatherData;
}
```

Comment (martynakr) [reply js/weatherApp.js:56]:
```js
try {
   const latLong = await getLatLong(input.value)
   const weatherData = await getWeatherData(latLong)
   someFunctionThtaCreatesWetherInfoAndAddsToDom(weatherData)
} catch(e) {
    alter("")
    return
}


```

File: index.html [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+                    </div>
+                    <button id="closeCalculatorModal">Close</button>
+                </div> 
+            </dialog>
+
+            
+
+
+
+
```
Comment (martynakr) [inline index.html:54]: Remove all the long spaces

**Review by martynakr — CHANGES_REQUESTED** [review]

**Review by obj809 — COMMENTED** [review]
These suggestions have been reviewed and integrated into main branch.

**Pull request merged**